### ISSUE TITLE
[java] ImplicitSwitchFallThrough: False negative when fall-through comment is placed after closing brace

### ISSUE THREAD

**Issue by Emilyaxe** [pmd/pmd#6629]
**Affects PMD Version:** 7.23.0

**Rule:** ImplicitSwitchFallThrough — https://docs.pmd-code.org/latest/pmd_rules_java_errorprone.html#implicitswitchfallthrough

**Description:**

`ImplicitSwitchFallThrough` does not fire when a `case` block contains braces without a `break`, and the trailing comment is placed after the closing brace. 

Although the control flow still falls through from `case 1` to `case 2`, PMD fails to detect it in this situation.

**Code Sample demonstrating the issue:**

```java
public class FallThroughDemo {
    // ImplicitSwitchFallThrough is NOT reported, but fall-through still occurs
    public static String processB(int opcode, String cst) {
        String result = cst;
        switch (opcode) {
            case 1: {
                result = result.toLowerCase(); // Line7:ImplicitSwitchFallThrough should be reported
            } // and fall through...
            case 2: {
                result = result.trim();
                break;
            }
            default: {
                result = "";
                break;
            }
        }
        return result;
    }
}
```

**Actual outcome:** PMD should report a violation at line 7, but doesn't. This is a false-negative.

**Running PMD through:** CLI

**Comment by UncleOwen** [comment]
If I remove the braces there is still no warning. So this is because of the comment!? Very interesting, thanks for the report!

**Comment by UncleOwen** [comment]
Oh, of course! PMD assumes that the comment is explaining the fall through. So this is kind of by design.

**Comment by Emilyaxe** [comment]
Thank you for the clarification!

I understand that PMD assumes the comment explains the fall-through behavior. However, I’m wondering whether this design is fully consistent in practice.

For example, in the following case:

```java
case 1: {
    result = result.toLowerCase();
    // and fall through...
}
```

ImplicitSwitchFallThrough is reported.

But in a semantically equivalent case:

```java
case 1: {
    result = result.toLowerCase();
} // and fall through...
```

no violation is reported.

Since the placement of the comment does not affect the actual control flow, it seems the analysis is sensitive to comment positioning rather than semantics. Would it make sense for the rule to handle these cases consistently?

**Comment by UncleOwen** [comment]
Right now, the comment has to be right before the case label. We could add a special case for allowing a `}` in between, but that kind of opens the door to looking in arbitrary places for the comment - which won't work.
Not saying yes, not saying no. Someone will have to think more about this, I'm just spitballing right now.

**Comment by Emilyaxe** [comment]
> Right now, the comment has to be right before the case label. We could add a special case for allowing a `}` in between, but that kind of opens the door to looking in arbitrary places for the comment - which won't work. Not saying yes, not saying no. Someone will have to think more about this, I'm just spitballing right now.

Thank you for the kind explanation.

**Comment by adangel** [comment]
As the rule documentation says:

> You can ignore a violation by commenting `// fallthrough` before the case label which is reached by fallthrough

I think, the comment should be really before the case label - that means, if the previous case uses a block, it must be outside of the block.

So, I'd say, this works as designed and we shouldn't change the behavior of the rule.

But: Can the documentation be improved to make this clearer?

Btw: the following words somewhere in the comment, make this violation being ignored: `fallthrough`, `fallsthrough`, `fall through`, `falls through`, `fall-through`, `falls-through`, `fallthru`, `fallsthru`, `fall-thru` - so it is not just `fallthrough` as in the doc.

Also: Has anyone checked, whether this case would be reported as a suppressed warning? IMHO this is just another way of suppressing this violation and should therefore be mentioned in the report. This would be a different issue, though.

**Comment by UncleOwen** [comment]
@adangel 

> Also: Has anyone checked, whether this case would be reported as a suppressed warning? IMHO this is just another way of suppressing this violation and should therefore be mentioned in the report.

Do we even have a mechanism for a rule reporting a suppression?

**Comment by zbynek** [comment]
> IMHO this is just another way of suppressing this violation

There are other rules that distinguish accidental and intentional usages of a construct, e.g. "unused variable" and "empty catch block" checks do not report variables/exceptions called `ignored`.  Maybe this is the same category and intentional fall-throughs indicated by a comment should not appear in the report at all, even as suppressed?

Making the regexp for comments configurable would allow users to effectively disable the comment feature, making `@SuppressWarnings` the only option to suppress this warning.
